File: frontend/src/types/order.ts
import type { Address, AddressCreatePayload } from "./account";

export interface ServiceOption {
  service_id: number;
  service_name: string;
  description?: string | null;
  base_price: number;
  estimated_days: number;
}

export interface PackageDetails {
  package_id: number;
  weight: number;
  height: number;
  length: number;
  width: number;
  package_type: string;
  fragile?: boolean | null;
  is_sealed?: boolean | null;
  inspection_policy?: string | null;
  security_level?: string | null;
  declared_value?: number | null;
}

export interface PackageDetailsCreatePayload {
  weight: number;
  height: number;
  length: number;
  width: number;
  package_type: string;
  fragile?: boolean;
  is_sealed?: boolean;
  inspection_policy?: string;
  security_level?: string;
  declared_value?: number;
}

export interface OrderSummary {
  order_id: string;
  customer_id: string;
  customer_name: string;
  service_name?: string | null;
  receiver_name?: string | null;
  destination?: string | null;
  total_price: number;
  order_status: string;
  created_at: string;
}

export interface ShipmentOrder {
  order_id: string;
  customer_id: string;
  service_id?: number | null;
  sender_address_id?: number | null;
  receiver_address_id?: number | null;
  package_id?: number | null;
  approved_by?: string | null;
  total_price: number;
  order_status: string;
  created_at: string;
  notes?: string | null;
}

export interface OrderDetailResponse {
  order: ShipmentOrder;
  service_option?: ServiceOption | null;
  sender_address?: Address | null;
  receiver_address?: Address | null;
  package_details?: PackageDetails | null;
}

export interface OrderFormState {
  sender_name: string;
  sender_phone: string;
  pickup_street: string;
  pickup_district: string;
  pickup_city: string;
  pickup_postal_code: string;
  receiver_name: string;
  receiver_phone: string;
  delivery_street: string;
  delivery_district: string;
  delivery_city: string;
  delivery_postal_code: string;
  weight: number;
  dimensions: string;
  package_type: string;
  declared_value: number;
  service_id: number;
  notes: string;
}

export interface OrderCreatePayload {
  customer_id: string;
  service_id: number;
  sender_address: AddressCreatePayload;
  receiver_address: AddressCreatePayload;
  package_details: PackageDetailsCreatePayload;
  notes?: string;
}

File: backend/app/schemas/order.py
from datetime import datetime
from typing import Literal
from pydantic import Field
from app.schemas.account import AddressCreate, AddressRead
from app.schemas.common import ORMModel
from app.schemas.service_option import ServiceOptionRead


PackageTypeEnum = Literal["Document", "Fragile", "Standard", "Heavy", "Cold/Fresh"]
InspectionPolicyEnum = Literal["No_Inspection", "Allow_Inspection", "Allow_Trial"]
SecurityLevelEnum = Literal["Standard", "Confidential", "High-Security"]


class PackageDetailsBase(ORMModel):
    weight: float = Field(..., gt=0, description="Weight in kg")
    height: float = Field(..., gt=0, description="Height in cm")
    length: float = Field(..., gt=0, description="Length in cm")
    width: float = Field(..., gt=0, description="Width in cm")
    package_type: PackageTypeEnum = "Standard"
    fragile: bool = False
    is_sealed: bool = False
    inspection_policy: InspectionPolicyEnum = "Allow_Inspection"
    security_level: SecurityLevelEnum = "Standard"
    declared_value: float = Field(default=0.0, ge=0)


class PackageDetailsCreate(PackageDetailsBase):
    pass


class PackageDetailsUpdate(ORMModel):
    weight: float | None = Field(default=None, gt=0)
    height: float | None = Field(default=None, gt=0)
    length: float | None = Field(default=None, gt=0)
    width: float | None = Field(default=None, gt=0)
    package_type: PackageTypeEnum | None = None
    fragile: bool | None = None
    is_sealed: bool | None = None
    inspection_policy: InspectionPolicyEnum | None = None
    security_level: SecurityLevelEnum | None = None
    declared_value: float | None = Field(default=None, ge=0)


class PackageDetailsRead(PackageDetailsBase):
    package_id: int


class ShipmentOrderBase(ORMModel):
    customer_id: str
    service_id: int | None = None
    sender_address_id: int | None = None
    receiver_address_id: int | None = None
    package_id: int | None = None
    approved_by: str | None = None
    total_price: float = Field(default=0.0, ge=0)
    order_status: str = "Pending"
    notes: str | None = None


class ShipmentOrderCreate(ORMModel):
    customer_id: str
    service_id: int
    
    # Support both existing address ID or new address object
    sender_address_id: int | None = None
    sender_address: AddressCreate | None = None
    
    receiver_address_id: int | None = None
    receiver_address: AddressCreate | None = None
    
    package_details: PackageDetailsCreate
    notes: str | None = None


class ShipmentOrderUpdate(ORMModel):
    service_id: int | None = None
    approved_by: str | None = None
    order_status: str | None = None
    notes: str | None = None


class ShipmentOrderRead(ShipmentOrderBase):
    order_id: str
    created_at: datetime


class ShipmentOrderListItem(ORMModel):
    order_id: str
    customer_id: str
    customer_name: str
    service_name: str | None = None
    receiver_name: str | None = None
    destination: str | None = None
    total_price: float
    order_status: str
    created_at: datetime


class ShipmentOrderResponse(ORMModel):
    order: ShipmentOrderRead
    service_option: ServiceOptionRead | None = None
    sender_address: AddressRead | None = None
    receiver_address: AddressRead | None = None
    package_details: PackageDetailsRead | None = None